lib/Dialect/Secret/Transforms/DistributeGeneric.cpp

```diff
@@ -0,0 +1,203 @@
+//     ^bb0(%used: i32):
+//       %0 = arith.muli %used, %used : i32
+//       secret.yield %0 : i32
+//    } -> (!secret.secret<i32>)
+//
+struct RemoveUnusedGenericArgs : public OpRewritePattern<GenericOp> {
+  RemoveUnusedGenericArgs(mlir::MLIRContext *context)
+      : OpRewritePattern<GenericOp>(context, /*benefit=*/1) {}
+
+  LogicalResult matchAndRewrite(GenericOp op,
+                                PatternRewriter &rewriter) const override {
+    bool hasUnusedOps = false;
+    Block *body = op.getBody();
+    for (BlockArgument arg : body->getArguments()) {
+      if (arg.use_empty()) {
+        hasUnusedOps = true;
+        rewriter.updateRootInPlace(op, [&]() {
+          body->eraseArgument(arg.getArgNumber());
+          op.getOperation()->eraseOperand(arg.getArgNumber());
+        });
+      }
+    }
+
+    return hasUnusedOps ? success() : failure();
+  }
+};
+
+// Split a secret.generic containing multiple ops into its underlying op if no
+// secret values are included in its input.
+//
+// E.g.,
+//
+//    %res = secret.generic ins(%value : !secret.secret<i32>) {
+//    ^bb0(%clear_value: i32):
+//      %c7 = arith.constant 7 : i32
+//      %0 = arith.muli %clear_value, %c7 : i32
+//      secret.yield %0 : i32
+//    } -> (!secret.secret<i32>)
+//
+// is transformed to
+//
+//    %secret_7 = secret.generic {
+//      %c7 = arith.constant 7 : i32
+//      secret.yield %c7 : i32
+//    } -> !secret.secret<i32>
+//    %1 = secret.generic ins(
+//       %arg0, %secret_7 : !secret.secret<i32>, !secret.secret<i32>) {
+//    ^bb0(%clear_arg0: i32, %clear_7: i32):
+//      %7 = arith.muli %clear_arg0, %clear_7 : i32
+//      secret.yield %7 : i32
+//    } -> !secret.secret<i32>
+//
+struct PeelFromGeneric : public OpRewritePattern<GenericOp> {
+  PeelFromGeneric(mlir::MLIRContext *context)
+      : OpRewritePattern<GenericOp>(context, /*benefit=*/1) {}
+
+  LogicalResult matchAndRewrite(GenericOp op,
+                                PatternRewriter &rewriter) const override {
+    Block *body = op.getBody();
+    unsigned numOps = body->getOperations().size();
+    if (numOps == 0) {
+      llvm_unreachable(
+          "secret.generic must have nonempty body (the verifier should enforce "
+          "this)");
+    }
+
+    // Recursive base case: stop if there's only one op left, noting that we
+    // check for 2 ops because the last op is enforced to be a yield op by the
+    // verifier.
+    if (numOps == 1 || numOps == 2) {
+      return failure();
+    }
+
+    Operation &firstOp = body->getOperations().front();
+    // The inputs to the op are generic op's block arguments (cleartext values),
+    // and they need to change to be the corresponding generic op's normal
+    // operands (maybe secret values).
+    SmallVector<Value> newInputs;
+    // The indices of the new inputs in the original block argument list
+    SmallVector<unsigned> newInputIndices;
+    for (Value val : firstOp.getOperands()) {
```

Comment:
do you need to check that `firstOp`'s operands don't use the secret inputs? (In the example case - that the firstOp does not use `clear_value`.)

Reply:
IMO this is something that should be enforced by the verifier: all ops within a secret.generic can only use as operands the values in the args of the containing basic block, and those args have to correspond to the cleartext versions of the secret.generic's operands in the same order. Added to https://github.com/google/heir/issues/108

Reply:
Oh, that part I understand - but my question is that it seems like from this description
```
// Split a secret.generic containing multiple ops into its underlying op if no
// secret values are included in its input
```

that this intends to peel apart the first op only when the op doesnt use the block arguments.

Now reading the code, it seems like maybe that comment is incorrect, and actually this is just breaking apart the secret.generic ops?

Reply:
Ah, yeah I think I had originally intended that, and then decided to split that into its own separate rewrite rule for simplicity. Will update the comment.